lib/vm-agent.js

```diff
@@ -231,6 +239,52 @@ VmAgent.prototype.updateVmapiVm = function updateVmapiVm(vmUuid, callback) {
             var opts = {log: self.log, uuid: vmUuid, include_dni: true};
             var startLoad = (new Date()).getTime();
 
+            function _handleLoadErr(_vmUuid, err, next) {
+                var loadErr;
+
+                if (err && err.restCode === 'VmNotFound') {
+                    if (self.knownDniVms[_vmUuid]) {
+                        // DNI VM was deleted, we don't send this to VMAPI but
+                        // we do want to purge it from the list after some delay
+                        // in case other delete events are emitted, we still
+                        // need to know that this was DNI.
+                        if (!self.knownDniVms[_vmUuid].timer) {
+                            self.knownDniVms[_vmUuid].timer = setTimeout(
+                                function _purgeDeletedDNI() {
+                                    self.log.debug({vmUuid: _vmUuid},
+                                        'purging DNI knowledge');
+                                    delete self.knownDniVms[_vmUuid];
+                                },
+                                DNI_PURGE_DELAY
+                            );
+                        }
+                        loadErr = new Error('deleted VM had do_not_inventory');
```

Comment:
It's an error the same way that VmNotFound is an error. We're being asked to update the VM, but we're saying that's not possible. It's also useful because an error allows us to early-exit from the vasync pipeline. If you have a better idea how all this should be handled, I'd be happy to hear it.